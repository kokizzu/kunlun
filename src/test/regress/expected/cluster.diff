308c308
<  clstr_tst            | r       | f
---
>  clstr_tst            | r       | t
314c314
<  clstr_tst_inh        | r       | f
---
>  clstr_tst_inh        | r       | t
419d418
<  1
424a424
>  1
