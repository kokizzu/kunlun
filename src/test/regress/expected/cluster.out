drop table if exists clstr_tst_s;
psql:sql/cluster.sql:1: NOTICE:  table "clstr_tst_s" does not exist, skipping
DROP TABLE
CREATE TABLE clstr_tst_s (rf_a SERIAL PRIMARY KEY,
	b INT);
CREATE TABLE
drop table if exists clstr_tst;
psql:sql/cluster.sql:5: NOTICE:  table "clstr_tst" does not exist, skipping
DROP TABLE
CREATE TABLE clstr_tst (a SERIAL PRIMARY KEY,
	b INT,
	c varchar(50),
	d text);
CREATE TABLE
CREATE INDEX clstr_tst_b ON clstr_tst (b);
CREATE INDEX
CREATE INDEX clstr_tst_c ON clstr_tst (c);
CREATE INDEX
CREATE INDEX clstr_tst_c_b ON clstr_tst (c,b);
CREATE INDEX
CREATE INDEX clstr_tst_b_c ON clstr_tst (b,c);
CREATE INDEX
INSERT INTO clstr_tst_s (b) VALUES (0);
INSERT 0 1
INSERT INTO clstr_tst_s (b) SELECT b FROM clstr_tst_s;
INSERT 0 1
INSERT INTO clstr_tst_s (b) SELECT b FROM clstr_tst_s;
INSERT 0 2
INSERT INTO clstr_tst_s (b) SELECT b FROM clstr_tst_s;
INSERT 0 4
INSERT INTO clstr_tst_s (b) SELECT b FROM clstr_tst_s;
INSERT 0 8
INSERT INTO clstr_tst_s (b) SELECT b FROM clstr_tst_s;
INSERT 0 16
drop table if exists clstr_tst_inh;
psql:sql/cluster.sql:23: NOTICE:  table "clstr_tst_inh" does not exist, skipping
DROP TABLE
CREATE TABLE clstr_tst_inh (like clstr_tst);
CREATE TABLE
INSERT INTO clstr_tst (b, c) VALUES (11, 'once');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (10, 'diez');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (31, 'treinta y uno');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (22, 'veintidos');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (3, 'tres');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (20, 'veinte');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (23, 'veintitres');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (21, 'veintiuno');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (4, 'cuatro');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (14, 'catorce');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (2, 'dos');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (18, 'dieciocho');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (27, 'veintisiete');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (25, 'veinticinco');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (13, 'trece');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (28, 'veintiocho');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (32, 'treinta y dos');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (5, 'cinco');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (29, 'veintinueve');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (1, 'uno');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (24, 'veinticuatro');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (30, 'treinta');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (12, 'doce');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (17, 'diecisiete');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (9, 'nueve');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (19, 'diecinueve');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (26, 'veintiseis');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (15, 'quince');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (7, 'siete');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (16, 'dieciseis');
INSERT 0 1
INSERT INTO clstr_tst (b, c) VALUES (8, 'ocho');
INSERT 0 1
-- This entry is needed to test that TOASTED values are copied correctly.
INSERT INTO clstr_tst (b, c, d) VALUES (6, 'seis', repeat('xyzzy', 100000));
INSERT 0 1
SELECT a,b,c,substring(d for 30), length(d) from clstr_tst;
 a  | b  |       c       |           substring            | length 
----+----+---------------+--------------------------------+--------
  1 | 11 | once          |                                |       
  2 | 10 | diez          |                                |       
  3 | 31 | treinta y uno |                                |       
  4 | 22 | veintidos     |                                |       
  5 |  3 | tres          |                                |       
  6 | 20 | veinte        |                                |       
  7 | 23 | veintitres    |                                |       
  8 | 21 | veintiuno     |                                |       
  9 |  4 | cuatro        |                                |       
 10 | 14 | catorce       |                                |       
 11 |  2 | dos           |                                |       
 12 | 18 | dieciocho     |                                |       
 13 | 27 | veintisiete   |                                |       
 14 | 25 | veinticinco   |                                |       
 15 | 13 | trece         |                                |       
 16 | 28 | veintiocho    |                                |       
 17 | 32 | treinta y dos |                                |       
 18 |  5 | cinco         |                                |       
 19 | 29 | veintinueve   |                                |       
 20 |  1 | uno           |                                |       
 21 | 24 | veinticuatro  |                                |       
 22 | 30 | treinta       |                                |       
 23 | 12 | doce          |                                |       
 24 | 17 | diecisiete    |                                |       
 25 |  9 | nueve         |                                |       
 26 | 19 | diecinueve    |                                |       
 27 | 26 | veintiseis    |                                |       
 28 | 15 | quince        |                                |       
 29 |  7 | siete         |                                |       
 30 | 16 | dieciseis     |                                |       
 31 |  8 | ocho          |                                |       
 32 |  6 | seis          | xyzzyxyzzyxyzzyxyzzyxyzzyxyzzy | 500000
(32 rows)

SELECT a,b,c,substring(d for 30), length(d) from clstr_tst ORDER BY a;
 a  | b  |       c       |           substring            | length 
----+----+---------------+--------------------------------+--------
  1 | 11 | once          |                                |       
  2 | 10 | diez          |                                |       
  3 | 31 | treinta y uno |                                |       
  4 | 22 | veintidos     |                                |       
  5 |  3 | tres          |                                |       
  6 | 20 | veinte        |                                |       
  7 | 23 | veintitres    |                                |       
  8 | 21 | veintiuno     |                                |       
  9 |  4 | cuatro        |                                |       
 10 | 14 | catorce       |                                |       
 11 |  2 | dos           |                                |       
 12 | 18 | dieciocho     |                                |       
 13 | 27 | veintisiete   |                                |       
 14 | 25 | veinticinco   |                                |       
 15 | 13 | trece         |                                |       
 16 | 28 | veintiocho    |                                |       
 17 | 32 | treinta y dos |                                |       
 18 |  5 | cinco         |                                |       
 19 | 29 | veintinueve   |                                |       
 20 |  1 | uno           |                                |       
 21 | 24 | veinticuatro  |                                |       
 22 | 30 | treinta       |                                |       
 23 | 12 | doce          |                                |       
 24 | 17 | diecisiete    |                                |       
 25 |  9 | nueve         |                                |       
 26 | 19 | diecinueve    |                                |       
 27 | 26 | veintiseis    |                                |       
 28 | 15 | quince        |                                |       
 29 |  7 | siete         |                                |       
 30 | 16 | dieciseis     |                                |       
 31 |  8 | ocho          |                                |       
 32 |  6 | seis          | xyzzyxyzzyxyzzyxyzzyxyzzyxyzzy | 500000
(32 rows)

SELECT a,b,c,substring(d for 30), length(d) from clstr_tst ORDER BY b;
 a  | b  |       c       |           substring            | length 
----+----+---------------+--------------------------------+--------
 20 |  1 | uno           |                                |       
 11 |  2 | dos           |                                |       
  5 |  3 | tres          |                                |       
  9 |  4 | cuatro        |                                |       
 18 |  5 | cinco         |                                |       
 32 |  6 | seis          | xyzzyxyzzyxyzzyxyzzyxyzzyxyzzy | 500000
 29 |  7 | siete         |                                |       
 31 |  8 | ocho          |                                |       
 25 |  9 | nueve         |                                |       
  2 | 10 | diez          |                                |       
  1 | 11 | once          |                                |       
 23 | 12 | doce          |                                |       
 15 | 13 | trece         |                                |       
 10 | 14 | catorce       |                                |       
 28 | 15 | quince        |                                |       
 30 | 16 | dieciseis     |                                |       
 24 | 17 | diecisiete    |                                |       
 12 | 18 | dieciocho     |                                |       
 26 | 19 | diecinueve    |                                |       
  6 | 20 | veinte        |                                |       
  8 | 21 | veintiuno     |                                |       
  4 | 22 | veintidos     |                                |       
  7 | 23 | veintitres    |                                |       
 21 | 24 | veinticuatro  |                                |       
 14 | 25 | veinticinco   |                                |       
 27 | 26 | veintiseis    |                                |       
 13 | 27 | veintisiete   |                                |       
 16 | 28 | veintiocho    |                                |       
 19 | 29 | veintinueve   |                                |       
 22 | 30 | treinta       |                                |       
  3 | 31 | treinta y uno |                                |       
 17 | 32 | treinta y dos |                                |       
(32 rows)

SELECT a,b,c,substring(d for 30), length(d) from clstr_tst ORDER BY c;
 a  | b  |       c       |           substring            | length 
----+----+---------------+--------------------------------+--------
 10 | 14 | catorce       |                                |       
 18 |  5 | cinco         |                                |       
  9 |  4 | cuatro        |                                |       
 26 | 19 | diecinueve    |                                |       
 12 | 18 | dieciocho     |                                |       
 30 | 16 | dieciseis     |                                |       
 24 | 17 | diecisiete    |                                |       
  2 | 10 | diez          |                                |       
 23 | 12 | doce          |                                |       
 11 |  2 | dos           |                                |       
 25 |  9 | nueve         |                                |       
 31 |  8 | ocho          |                                |       
  1 | 11 | once          |                                |       
 28 | 15 | quince        |                                |       
 32 |  6 | seis          | xyzzyxyzzyxyzzyxyzzyxyzzyxyzzy | 500000
 29 |  7 | siete         |                                |       
 15 | 13 | trece         |                                |       
 22 | 30 | treinta       |                                |       
 17 | 32 | treinta y dos |                                |       
  3 | 31 | treinta y uno |                                |       
  5 |  3 | tres          |                                |       
 20 |  1 | uno           |                                |       
  6 | 20 | veinte        |                                |       
 14 | 25 | veinticinco   |                                |       
 21 | 24 | veinticuatro  |                                |       
  4 | 22 | veintidos     |                                |       
 19 | 29 | veintinueve   |                                |       
 16 | 28 | veintiocho    |                                |       
 27 | 26 | veintiseis    |                                |       
 13 | 27 | veintisiete   |                                |       
  7 | 23 | veintitres    |                                |       
  8 | 21 | veintiuno     |                                |       
(32 rows)

-- Verify that inheritance link still works
INSERT INTO clstr_tst_inh VALUES (0, 100, 'in child table');
INSERT 0 1
SELECT a,b,c,substring(d for 30), length(d) from clstr_tst;
 a  | b  |       c       |           substring            | length 
----+----+---------------+--------------------------------+--------
  1 | 11 | once          |                                |       
  2 | 10 | diez          |                                |       
  3 | 31 | treinta y uno |                                |       
  4 | 22 | veintidos     |                                |       
  5 |  3 | tres          |                                |       
  6 | 20 | veinte        |                                |       
  7 | 23 | veintitres    |                                |       
  8 | 21 | veintiuno     |                                |       
  9 |  4 | cuatro        |                                |       
 10 | 14 | catorce       |                                |       
 11 |  2 | dos           |                                |       
 12 | 18 | dieciocho     |                                |       
 13 | 27 | veintisiete   |                                |       
 14 | 25 | veinticinco   |                                |       
 15 | 13 | trece         |                                |       
 16 | 28 | veintiocho    |                                |       
 17 | 32 | treinta y dos |                                |       
 18 |  5 | cinco         |                                |       
 19 | 29 | veintinueve   |                                |       
 20 |  1 | uno           |                                |       
 21 | 24 | veinticuatro  |                                |       
 22 | 30 | treinta       |                                |       
 23 | 12 | doce          |                                |       
 24 | 17 | diecisiete    |                                |       
 25 |  9 | nueve         |                                |       
 26 | 19 | diecinueve    |                                |       
 27 | 26 | veintiseis    |                                |       
 28 | 15 | quince        |                                |       
 29 |  7 | siete         |                                |       
 30 | 16 | dieciseis     |                                |       
 31 |  8 | ocho          |                                |       
 32 |  6 | seis          | xyzzyxyzzyxyzzyxyzzyxyzzyxyzzy | 500000
(32 rows)

-- Verify that foreign key link still works
INSERT INTO clstr_tst (b, c) VALUES (1111, 'this should fail');
INSERT 0 1
SELECT conname FROM pg_constraint WHERE conrelid = 'clstr_tst'::regclass
ORDER BY 1;
    conname     
----------------
 clstr_tst_pkey
(1 row)

SELECT relname, relkind,
    EXISTS(SELECT 1 FROM pg_class WHERE oid = c.reltoastrelid) AS hastoast
FROM pg_class c WHERE relname LIKE 'clstr_tst%' ORDER BY relname;
       relname        | relkind | hastoast 
----------------------+---------+----------
 clstr_tst            | r       | f
 clstr_tst_a_seq      | S       | f
 clstr_tst_b          | i       | f
 clstr_tst_b_c        | i       | f
 clstr_tst_c          | i       | f
 clstr_tst_c_b        | i       | f
 clstr_tst_inh        | r       | f
 clstr_tst_pkey       | i       | f
 clstr_tst_s          | r       | f
 clstr_tst_s_pkey     | i       | f
 clstr_tst_s_rf_a_seq | S       | f
(11 rows)

-- Verify that indisclustered is correctly set
SELECT pg_class.relname FROM pg_index, pg_class, pg_class AS pg_class_2
WHERE pg_class.oid=indexrelid
	AND indrelid=pg_class_2.oid
	AND pg_class_2.relname = 'clstr_tst'
	AND indisclustered;
 relname 
---------
(0 rows)

-- Try changing indisclustered
SELECT pg_class.relname FROM pg_index, pg_class, pg_class AS pg_class_2
WHERE pg_class.oid=indexrelid
	AND indrelid=pg_class_2.oid
	AND pg_class_2.relname = 'clstr_tst'
	AND indisclustered;
 relname 
---------
(0 rows)

-- Try turning off all clustering
SELECT pg_class.relname FROM pg_index, pg_class, pg_class AS pg_class_2
WHERE pg_class.oid=indexrelid
	AND indrelid=pg_class_2.oid
	AND pg_class_2.relname = 'clstr_tst'
	AND indisclustered;
 relname 
---------
(0 rows)

-- Verify that clustering all tables does in fact cluster the right ones
CREATE USER regress_clstr_user;
CREATE ROLE
CREATE TABLE clstr_1 (a INT PRIMARY KEY);
CREATE TABLE
CREATE TABLE clstr_2 (a INT PRIMARY KEY);
CREATE TABLE
CREATE TABLE clstr_3 (a INT PRIMARY KEY);
CREATE TABLE
ALTER TABLE clstr_1 OWNER TO regress_clstr_user;
ALTER TABLE
ALTER TABLE clstr_3 OWNER TO regress_clstr_user;
ALTER TABLE
GRANT SELECT ON clstr_2 TO regress_clstr_user;
GRANT
INSERT INTO clstr_1 VALUES (2);
INSERT 0 1
INSERT INTO clstr_1 VALUES (1);
INSERT 0 1
INSERT INTO clstr_2 VALUES (2);
INSERT 0 1
INSERT INTO clstr_2 VALUES (1);
INSERT 0 1
INSERT INTO clstr_3 VALUES (2);
INSERT 0 1
INSERT INTO clstr_3 VALUES (1);
INSERT 0 1
SELECT * FROM clstr_1 UNION ALL
  SELECT * FROM clstr_2 UNION ALL
  SELECT * FROM clstr_3 order by 1;
 a 
---
 1
 1
 1
 2
 2
 2
(6 rows)

-- revert to the original state
DELETE FROM clstr_1;
DELETE 2
DELETE FROM clstr_2;
DELETE 2
DELETE FROM clstr_3;
DELETE 2
INSERT INTO clstr_1 VALUES (2);
INSERT 0 1
INSERT INTO clstr_1 VALUES (1);
INSERT 0 1
INSERT INTO clstr_2 VALUES (2);
INSERT 0 1
INSERT INTO clstr_2 VALUES (1);
INSERT 0 1
INSERT INTO clstr_3 VALUES (2);
INSERT 0 1
INSERT INTO clstr_3 VALUES (1);
INSERT 0 1
-- this user can only cluster clstr_1 and clstr_3, but the latter
-- has not been clustered
SET SESSION AUTHORIZATION regress_clstr_user;
SET
SELECT * FROM clstr_1 UNION ALL
  SELECT * FROM clstr_2 UNION ALL
  SELECT * FROM clstr_3;
 a 
---
 1
 2
 1
 2
 1
 2
(6 rows)

-- cluster a single table using the indisclustered bit previously set
DELETE FROM clstr_1;
DELETE 2
INSERT INTO clstr_1 VALUES (2);
INSERT 0 1
INSERT INTO clstr_1 VALUES (1);
INSERT 0 1
SELECT * FROM clstr_1 order by 1;
 a 
---
 1
 2
(2 rows)

-- Test MVCC-safety of cluster. There isn't much we can do to verify the
-- results with a single backend...
CREATE TABLE clustertest (key1 int PRIMARY KEY);
CREATE TABLE
INSERT INTO clustertest VALUES (10);
INSERT 0 1
INSERT INTO clustertest VALUES (20);
INSERT 0 1
INSERT INTO clustertest VALUES (30);
INSERT 0 1
INSERT INTO clustertest VALUES (40);
INSERT 0 1
INSERT INTO clustertest VALUES (50);
INSERT 0 1
BEGIN;
BEGIN
-- Test update where the old row version is found first in the scan
UPDATE clustertest SET key1 = 100 WHERE key1 = 10;
UPDATE 1
-- Test update where the new row version is found first in the scan
UPDATE clustertest SET key1 = 35 WHERE key1 = 40;
UPDATE 1
-- Test longer update chain
UPDATE clustertest SET key1 = 60 WHERE key1 = 50;
UPDATE 1
UPDATE clustertest SET key1 = 70 WHERE key1 = 60;
UPDATE 1
UPDATE clustertest SET key1 = 80 WHERE key1 = 70;
UPDATE 1
SELECT * FROM clustertest order by key1;
 key1 
------
   20
   30
   35
   80
  100
(5 rows)

SELECT * FROM clustertest order by key1;
 key1 
------
   20
   30
   35
   80
  100
(5 rows)

COMMIT;
COMMIT
SELECT * FROM clustertest order by key1;
 key1 
------
   20
   30
   35
   80
  100
(5 rows)

-- check that temp tables can be clustered
create temp table clstr_temp (col1 int primary key, col2 text);
CREATE TABLE
insert into clstr_temp values (2, 'two'), (1, 'one');
INSERT 0 2
select * from clstr_temp;
 col1 | col2 
------+------
    2 | two
    1 | one
(2 rows)

drop table clstr_temp;
DROP TABLE
RESET SESSION AUTHORIZATION;
RESET
-- Check that partitioned tables cannot be clustered
CREATE TABLE clstrpart (a int) PARTITION BY RANGE (a);
CREATE TABLE
CREATE INDEX clstrpart_idx ON clstrpart (a);
CREATE INDEX
DROP TABLE clstrpart;
DROP TABLE
create table clstr_4 (like tenk1);
CREATE TABLE
insert into clstr_4 select * from tenk1;
INSERT 0 0
create index cluster_sort on clstr_4 (hundred, thousand, tenthous);
CREATE INDEX
set enable_indexscan = off;
SET
-- Use external sort:
set maintenance_work_mem = '1MB';
SET
select * from
(select hundred, lag(hundred) over () as lhundred,
        thousand, lag(thousand) over () as lthousand,
        tenthous, lag(tenthous) over () as ltenthous from clstr_4) ss
where row(hundred, thousand, tenthous) <= row(lhundred, lthousand, ltenthous);
 hundred | lhundred | thousand | lthousand | tenthous | ltenthous 
---------+----------+----------+-----------+----------+-----------
(0 rows)

reset enable_indexscan;
RESET
reset maintenance_work_mem;
RESET
-- clean up
DROP TABLE clustertest;
DROP TABLE
DROP TABLE clstr_1;
DROP TABLE
DROP TABLE clstr_2;
DROP TABLE
DROP TABLE clstr_3;
DROP TABLE
DROP TABLE clstr_4;
DROP TABLE
DROP USER regress_clstr_user;
DROP ROLE
